// @zenfs/dom/dist/devices/framebuffer.js
import { Errno, ErrnoError } from '@zenfs/core';
let framebufferN = 0;
/**
 * A frame buffer
 *
 * Setup:
 * ```
 * addDevice(framebuffer, { canvas: document.querySelector('#your-canvas') })
 * ```
 */
export const framebuffer = {
    name: 'framebuffer',
    init(ino, { canvas } = {}) {
        if (!canvas) {
            canvas = document.createElement('canvas');
            document.body.appendChild(canvas);
        }
        const context = canvas.getContext('2d');
        if (!context) {
            throw new ErrnoError(Errno.EIO, 'Could not get context from canvas whilst initializing frame buffer.');
        }
        const image = new ImageData(canvas.width, canvas.height);
        return {
            data: { context, image },
            major: 29,
            minor: framebufferN++,
            name: 'fb',
        };
    },
    read() { },
    write({ data: { image, context } }, buffer, offset) {
        image.data.set(buffer, offset);
        context.putImageData(image, 0, 0);
        return buffer.byteLength;
    },
};
